core/src/credential.rs

```diff
@@ -0,0 +1,143 @@
+// trait X allows us to encode that object into JSON-LD
+// We provide types that implement trait X for the cred types that we support
+// Users can also user their own types that implement trait X if they need a different structure
+// --- 
+// Default context and Cred types are defaulted but can be redefined 
+
+type VerificationContext = [&'static str];
+
+pub(crate) const JSON_LD_CONTEXT:  VerificationContext = ["https://www.w3.org/2018/credentials/v1", "https://www.w3.org/2018/credentials/examples/v1"];
+
+pub const CRED_TYPE_PERMANENT_RESIDENT_CARD: str = "PermanentResidentCard";
+pub const CRED_TYPE_BANK_CARD: str = "BankCard";
+
+type CredentialSubject = serde_json::Value;
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct Credential <'a> {
+    #[serde(flatten)]
+    verifiable_credential:VerifiableCredential <'a>,
+    #[serde(flatten)]
+    proof: CredentialProof <'a>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct VerifiableCredential <'a> {
+    #[serde(rename = "@context")]
+	context:  VerificationContext,
+    #[serde(rename = "@id")]
+	id: String,
+    cred_type:  String,
+	issuer: String,
+	issuance_date: SystemTime,
+	subject: CredentialSubject,
+}
+
+pub struct CredentialProof <'a> {
+	proof_type: String,
+	created: String,
+	verification_method: String,
+	proof_purpose: String,
+	proof_value: String,
+}
+
+pub struct CredentialManager {
+    credential_subject: HashMap<String, String>,
+    context: VerificationContext,
+}
+
+trait VC {
+    fn getSubject(&self, key: &str) -> Result<String, Box<dyn std::error::Error>> {
+        match self.credential_subject.get(key) {
+            Some(&value) => Ok(value.to_string()),
+            None => Err(Error::new(ErrorKind::Other, "Key not found")),
+        } 
+    }
+    fn getContext(&self) -> VerificationContext {
+        self.context.clone()
+    }
+    fn getCredentialTypes(&self) -> [String] {
+        [CRED_TYPE_PERMANENT_RESIDENT_CARD, CRED_TYPE_BANK_CARD];
+    }
+}
+
+/// Default VC trait implementation for CredentialManager.
+/// Callers may reimplement this trait in their code by: impl VC for CredentialManager {/...code here.../}
+impl VC for CredentialManager {}
+
+impl CredentialManager {
+    /// returns CredentialManager cred_subject_template with default configuration accessible for overlapping via VC trait
+    pub fn new() -> CredentialManager {
+        let cred_subject_template: HashMap<&str, &str> = HashMap::from([
+                (CRED_TYPE_PERMANENT_RESIDENT_CARD, r###"{
+                    "@id": "https://w3id.org/citizenship#PermanentResidentCard",
+                    "@context": {
+                      "@version": 1.1,
+                      "@protected": true,
+                      "id": "@id",
+                      "type": "[
+                        "PermanentResidentCard"
+                     ]",
+                      "description": "http://schema.org/description",
```

Comment:
Please note description, identifier, name are outside of the credential subject and belong in a level up.  Also please check these values- they have to be passed in by the user, and are not these predefined string values
  "id": "https://issuer.oidp.uscis.gov/credentials/83627465",
  "type": ["VerifiableCredential", "PermanentResidentCard"],
  "issuer": "did:example:28394728934792387",
  "identifier": "83627465",
  "name": "Permanent Resident Card",
  "description": "Government of Example Permanent Resident Card.",
  "issuanceDate": "2019-12-03T12:19:52Z",
  "expirationDate": "2029-12-03T12:19:52Z",
  "credentialSubject": {
    "id": "did:example:b34ca6cd37bbf23",
    "type": ["PermanentResident", "Person"],
    "givenName": "JOHN",
    "familyName": "SMITH",
    "gender": "Male",
    "image": "data:image/png;base64,iVBORw0KGgo...kJggg==",
    "residentSince": "2015-01-01",
    "lprCategory": "C09",
    "lprNumber": "999-999-999",
    "commuterClassification": "C1",
    "birthCountry": "Bahamas",
    "birthDate": "1958-07-17"
  },